File: 03_avl_tree/C_C++/avl_tree.c
#include<stdio.h>
#include<stdlib.h>
#include "avl_tree.h"

int get_tree_height(struct Node *node) {
	if (node == NULL) {
		return 0;
	}

	return node->height;
}

int get_max_value(int a, int b) {
	if (a > b) {
		return a;
	}

	return b;
}

int get_balance_factor(struct Node *node) {
	if (node == NULL) {
		return 0;
	}

	return (get_tree_height(node->left) - get_tree_height(node->right));
}

struct Node *rotate_right(struct Node *node) {
	struct Node *left_side = node->left;
	struct Node *tmp = left_side->right;

	left_side->right = node;
	node->left = tmp;

	node->height = get_max_value(get_tree_height(node->left), get_tree_height(node->right))+1;
	left_side->height = get_max_value(get_tree_height(left_side->left), get_tree_height(left_side->right))+1;

	return left_side;
}

struct Node *rotate_left(struct Node *node) {
	struct Node *right_side = node->right;
	struct Node *tmp = right_side->left;

	right_side->left = node;
	node->right = tmp;

	node->height = get_max_value(get_tree_height(node->left), get_tree_height(node->right))+1;
	right_side->height = get_max_value(get_tree_height(right_side->left), get_tree_height(right_side->right))+1;

	return right_side;
}

struct Node* new_node(char data) {
	struct Node* node = (struct Node*) malloc(sizeof(struct Node));
	node->data = data;
	node->left = NULL;
	node->right = NULL;
	
	return node;
}

struct Node* insert(struct Node* node, char data) {
	if (node == NULL) {
		return new_node(data);
	}

	if (data < node->data) {
		node->left = insert(node->left, data);
	} else if (data > node->data) {
		node->right = insert(node->right, data);
	} else {
		return node;
	}

	node->height = 1 + get_max_value(get_tree_height(node->left), get_tree_height(node->right));
	int balance = get_balance_factor(node);

	if (balance > 1 && data < node->left->data) {
		return rotate_right(node);
	}

	if (balance < -1 && data > node->right->data) {
		return rotate_left(node);
	}

	if (balance > 1 && data > node->left->data) {
		node->left = rotate_left(node->left);
		return rotate_right(node);
	}

	if (balance < -1 && data < node->right->data) {
		node->right = rotate_right(node->right);
		return rotate_left(node);
	}

	return node;
}

struct Node * min_node_value(struct Node* node) { 
	struct Node* current = node;

	while (current->left != NULL) {
		current = current->left;
	}

	return current; 
}
 
struct Node* delete_node(struct Node* root, char key) {
	if (root == NULL) { 
		return NULL;
	}

	if (key < root->data) { 
		root->left = delete_node(root->left, key);
	} else if( key > root->data ) { 
		root->right = delete_node(root->right, key);
	} else {
		if((root->left == NULL) || (root->right == NULL)) {
			struct Node *temp = root->left ? root->left : root->right; 

			if (temp == NULL) { 
				temp = root;
				root = NULL;
			} else {
				*root = *temp;
			}

			free(temp); 
		} else {
			struct Node* temp = min_node_value(root->right);
			root->data = temp->data;
			root->right = delete_node(root->right, temp->data); 
		}
	}

	if (root == NULL) {
		return NULL;
	}

	root->height = 1 + get_max_value(get_tree_height(root->left), get_tree_height(root->right));
	int balance = get_balance_factor(root); 

	if (balance > 1 && get_balance_factor(root->left) >= 0) {
		return rotate_right(root);
	}

	if (balance > 1 && get_balance_factor(root->left) < 0) {
		root->left =  rotate_left(root->left);
		return rotate_right(root);
	}

	if (balance < -1 && get_balance_factor(root->right) <= 0) {
		return rotate_left(root);
	}

	if (balance < -1 && get_balance_factor(root->right) > 0) {
		root->right = rotate_right(root->right);
		return rotate_left(root);
	}

	return root;
}

void pre_order(struct Node *root) {
	if(root != NULL) {
		printf("data: %c, height: %d\n", root->data, root->height);
		pre_order(root->left);
		pre_order(root->right);
	}
}

void in_order(struct Node *root) {
	if (root != NULL) {
		in_order(root->left);
		printf("data: %c, height: %d\n", root->data, root->height);
		in_order(root->right);
	}
}

void post_order(struct Node *root) {
	if (root != NULL) {
		post_order(root->left);
		post_order(root->right);
		printf("data: %c, height: %d\n", root->data, root->height);
	}
}

void clear_tree(struct Node *node) {
	if (node != NULL) {
		clear_tree(node->left);
		clear_tree(node->right);
		free(node);
	}
}

File: 03_avl_tree/python/AvlTree.py
#!/usr/bin/python3

class AvlTree(object):
	data = None
	left = None
	right = None
	height = 0

	def __init__(self, data):
		self.data = data
		self.left = None
		self.right = None
		self.height = 1
	# end constructor

	def insert(self, root, key):
		if root is None:
			return AvlTree(key) 
		elif key < root.data: 
			root.left = self.insert(root.left, key)
		else: 
			root.right = self.insert(root.right, key)
		# end if

		root.height = 1 + self.getMaxValue(self.getHeight(root.left), self.getHeight(root.right))
		bf = self.getBalance(root)

		if bf > 1 and key < root.left.data:
			return self.rightRotate(root)
		# end if

		if bf < -1 and key > root.right.data:
			return self.leftRotate(root)
		# end if

		if bf > 1 and key > root.left.data:
			root.left = self.leftRotate(root.left)
			return self.rightRotate(root)
		# end if

		if bf < -1 and key < root.right.data:
			root.right = self.rightRotate(root.right)
			return self.leftRotate(root)
		# end if

		return root
	# end method

	def delete(self, root, key):
		if root is None: 
			return  None
		# end if

		if key < root.data:
			root.left = self.delete(root.left, key)
		elif key > root.data:
			root.right = self.delete(root.right, key)
		else:
			if root.left is None:
				temp = root.right
				root = None
				return temp
			elif root.right is None:
				temp = root.left
				root = None
				return temp
			# end if

			temp = self.getMinValueNode(root.right)
			root.data = temp.data
			root.right = self.delete(root.right, temp.data)
		# end if

		if root is None:
			return None
		# end if

		root.height = 1 + self.getMaxValue(self.getHeight(root.left), self.getHeight(root.right))
		bf = self.getBalance(root)

		if bf > 1 and self.getBalance(root.left) >= 0:
			return self.rightRotate(root)
		# end if

		if bf < -1 and self.getBalance(root.right) <= 0:
			return self.leftRotate(root)
		# end if 

		if bf > 1 and self.getBalance(root.left) < 0:
			root.left = self.leftRotate(root.left)
			return self.rightRotate(root)
		# end if

		if bf < -1 and self.getBalance(root.right) > 0:
			root.right = self.rightRotate(root.right)
			return self.leftRotate(root)
		# end if

		return root
	# end method

	def leftRotate(self, node):
		rightHand = node.right
		tmp = rightHand.left

		rightHand.left = node
		node.right = tmp

		node.height = 1 + self.getMaxValue(self.getHeight(node.left), self.getHeight(node.right))
		rightHand.height = 1 + self.getMaxValue(self.getHeight(rightHand.left), self.getHeight(rightHand.right))

		return rightHand
	# end method

	def rightRotate(self, node):
		leftHand = node.left
		tmp = leftHand.right

		leftHand.right = node
		node.left = tmp

		node.height = 1 + self.getMaxValue(self.getHeight(node.left), self.getHeight(node.right))
		leftHand.height = 1 + self.getMaxValue(self.getHeight(leftHand.left), self.getHeight(leftHand.right))

		return leftHand
	# end method

	def getHeight(self, root):
		if root is None:
			return 0
		# end if

		return root.height
	# end method

	def getBalance(self, root):
		if root is None:
			return 0
		# end if

		return (self.getHeight(root.left) - self.getHeight(root.right))
	# end method

	def getMinValueNode(self, root):
		current = root

		while current.left is not None:
			current = current.left
		# end while

		return current
	# end method

	def preOrder(self, node):
		if node is not None:
			leftSide = None
			rightSide = None

			if node.left is not None:
				leftSide = node.left.data
			# end if

			if node.right is not None:
				rightSide = node.right.data
			# end if

			print(" data: {0} (height: {1}, left: {2}, right: {3})".format(node.data, node.height, leftSide, rightSide)) 
			self.preOrder(node.left)
			self.preOrder(node.right)
		# end if
	# end method

	def inOrder(self, node):
		if node is not None:
			self.inOrder(node.left)

			leftSide = None
			rightSide = None

			if node.left is not None:
				leftSide = node.left.data
			# end if

			if node.right is not None:
				rightSide = node.right.data
			# end if

			print(" data: {0} (height: {1}, left: {2}, right: {3})".format(node.data, node.height, leftSide, rightSide)) 
			self.inOrder(node.right)
		# end if
	# end method

	def postOrder(self, node):
		if node is not None:
			self.postOrder(node.left)
			self.postOrder(node.right)

			leftSide = None
			rightSide = None

			if node.left is not None:
				leftSide = node.left.data
			# end if

			if node.right is not None:
				rightSide = node.right.data
			# end if

			print(" data: {0} (height: {1}, left: {2}, right: {3})".format(node.data, node.height, leftSide, rightSide))
		# end if
	# end method

	def getMaxValue(self, a, b):
		if a > b:
			return a
		# end if

		return b
	#end method

	def containsNode(self, node, key):
		if node is not None:
			current = node

			while current is not None:
				if current.data > key:
					current = current.left
				elif current.data < key:
					current = current.right
				else:
					break
				# end if
			# end while

			if current is not None:
				return True
			# end if
		# end if

		return False
	# end method

	def clearTree(self, node):
		if node is not None:
			self.clearTree(node.left)
			self.clearTree(node.right)
			del node
		# end if
	# end method
# end class